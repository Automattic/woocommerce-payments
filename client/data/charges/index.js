/** @format */

/**
 * Internal dependencies
 */
import reducer from './reducer.gen';
import * as selectors from './selectors.gen';
import * as actions from './actions.gen';
import * as resolvers from './resolvers';

export { reducer, selectors, actions, resolvers };
export * from './hooks.gen';
