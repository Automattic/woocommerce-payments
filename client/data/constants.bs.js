// Generated by BUCKLESCRIPT, PLEASE EDIT WITH CARE


var namespace = "/wc/v3/payments";

var storeName = "wc/payments";

export {
  namespace ,
  storeName ,
  
}
/* No side effect */
