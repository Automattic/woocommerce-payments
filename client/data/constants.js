/** @format */

export const NAMESPACE = '/wc/v3/payments';
export const STORE_NAME = 'wc/payments';

export const ID_PREFIX = {
	transactions: 'transactions',
};
