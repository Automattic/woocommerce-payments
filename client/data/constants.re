let namespace = "/wc/v3/payments";
let storeName = "wc/payments";
