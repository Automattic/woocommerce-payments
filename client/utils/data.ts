/** @format */

export const getResourceId = (
	identifier: Record< string, string >
): string => {
	return JSON.stringify( identifier, Object.keys( identifier ).sort() );
};
