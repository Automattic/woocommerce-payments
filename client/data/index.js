/** @format */

/**
 * Internal dependencies
 */
import { STORE_NAME } from './constants';
import { initStore } from './store';

initStore();

export const WCPAY_STORE_NAME = STORE_NAME;

// We only ask for hooks when importing directly from 'data'.
export * from './deposits/hooks';
export * from './transactions/hooks';
export * from './charges/hooks';
export * from './timeline/hooks';
export * from './disputes/hooks';
export * from './settings/hooks';
export * from './multi-currency';
