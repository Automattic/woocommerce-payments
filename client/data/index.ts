/** @format */

/**
 * Internal dependencies
 */
import { STORE_NAME } from './constants';
import { initStore } from './store';

initStore();

// eslint-disable-next-line @typescript-eslint/naming-convention
export const WCPAY_STORE_NAME = STORE_NAME;

// We only ask for hooks when importing directly from 'wcpay/data'.
export * from './deposits/hooks';
export * from './transactions/hooks';
export * from './charges/hooks';
export * from './timeline/hooks';
export * from './disputes/hooks';
export * from './settings/hooks';
export * from './multi-currency';
export * from './card-readers/hooks';
export * from './capital/hooks';
export * from './documents/hooks';
export * from './payment-intents/hooks';
export * from './authorizations/hooks';
