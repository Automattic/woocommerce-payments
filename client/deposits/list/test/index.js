/** @format */

/**
 * External dependencies
 */
import { render } from '@testing-library/react';
import { updateQueryString } from '@woocommerce/navigation';
import { downloadCSVFile } from '@woocommerce/csv-export';
import { dateI18n } from '@wordpress/date';
import moment from 'moment';
import os from 'os';

/**
 * Internal dependencies
 */
import { DepositsList } from '../';
import { useDeposits, useDepositsSummary } from 'wcpay/data';

jest.mock( 'wcpay/data', () => ( {
	useDeposits: jest.fn(),
	useDepositsSummary: jest.fn(),
} ) );

jest.mock( '@woocommerce/csv-export', () => {
	const actualModule = jest.requireActual( '@woocommerce/csv-export' );

	return {
		...actualModule,
		downloadCSVFile: jest.fn(),
	};
} );

const mockDeposits = [
	{
		id: 'po_mock1',
		date: '2020-01-02 17:46:02',
		type: 'deposit',
		amount: 2000,
		status: 'paid',
		bankAccount: 'MOCK BANK •••• 1234 (USD)',
	},
	{
		id: 'po_mock2',
		date: '2020-01-03 17:46:02',
		type: 'withdrawal',
		amount: 3000,
		status: 'pending',
		bankAccount: 'MOCK BANK •••• 1234 (USD)',
	},
];

describe( 'Deposits list', () => {
	beforeEach( () => {
		jest.clearAllMocks();

		// the query string is preserved across tests, so we need to reset it
		updateQueryString( {}, '/', {} );

		global.wcpaySettings = { zeroDecimalCurrencies: [] };
	} );

	// this also covers structural test for single currency.
	test( 'renders correctly with multiple currencies', () => {
		useDeposits.mockReturnValue( {
			deposits: mockDeposits,
			depositsCount: 2,
			isLoading: false,
		} );

		useDepositsSummary.mockReturnValue( {
			depositsSummary: {
				count: 2,
				total: 5000,
				store_currencies: [ 'usd', 'eur' ],
			},
			isLoading: false,
		} );

		const { container } = render( <DepositsList /> );
		expect( container ).toMatchSnapshot();
	} );

	test( 'renders correctly a single deposit', () => {
		useDeposits.mockReturnValue( {
			deposits: mockDeposits,
			depositsCount: 1,
			isLoading: false,
		} );

		useDepositsSummary.mockReturnValue( {
			depositsSummary: {
				count: 1,
				total: 5000,
				store_currencies: [ 'usd' ],
			},
			isLoading: false,
		} );

		const { container } = render( <DepositsList /> );
		expect( container ).toMatchSnapshot();
	} );

	test( 'renders table summary only when the deposits summary data is available', () => {
		useDeposits.mockReturnValue( {
			deposits: mockDeposits,
			isLoading: false,
		} );

		useDepositsSummary.mockReturnValue( {
			depositsSummary: {
				count: 30,
			},
			isLoading: true,
		} );

		let { container } = render( <DepositsList /> );
		let tableSummary = container.querySelectorAll(
			'.woocommerce-table__summary'
		);

		expect( tableSummary ).toHaveLength( 0 );

		useDepositsSummary.mockReturnValue( {
			depositsSummary: {
				count: 2,
				total: 100,
			},
			isLoading: false,
		} );

		( { container } = render( <DepositsList /> ) );
		tableSummary = container.querySelectorAll(
			'.woocommerce-table__summary'
		);

		expect( tableSummary ).toHaveLength( 1 );
	} );

	describe( 'Download button', () => {
		test( 'renders when there are one or more deposits', () => {
			useDeposits.mockReturnValue( {
				deposits: mockDeposits,
				isLoading: false,
			} );

			const { queryByRole } = render( <DepositsList /> );
			const button = queryByRole( 'button', { name: 'Download' } );

			expect( button ).not.toBeNull();
		} );

		test( 'does not render when there are no deposits', () => {
			useDeposits.mockReturnValue( {
				deposits: [],
				isLoading: false,
			} );

			const { queryByRole } = render( <DepositsList /> );
			const button = queryByRole( 'button', { name: 'Download' } );

			expect( button ).toBeNull();
		} );
	} );

	describe( 'CSV download', () => {
		beforeEach( () => {
			useDeposits.mockReturnValue( {
				deposits: mockDeposits,
				depositsCount: 2,
				isLoading: false,
			} );
			useDepositsSummary.mockReturnValue( {
				depositsSummary: {
					count: 2,
					total: 5000,
				},
				isLoading: false,
			} );
		} );

		afterEach( () => {
			jest.resetAllMocks();
		} );

		afterAll( () => {
			jest.restoreAllMocks();
		} );

		test( 'should render expected columns in CSV when the download button is clicked', () => {
			const { getByRole } = render( <DepositsList /> );
			getByRole( 'button', { name: 'Download' } ).click();

			const expected = [
				'"Deposit Id"',
				'Date',
				'Type',
				'Amount',
				'Status',
				'"Bank account"',
			];

			const csvContent = downloadCSVFile.mock.calls[ 0 ][ 1 ];
			const csvHeaderRow = csvContent.split( os.EOL )[ 0 ].split( ',' );
			expect( csvHeaderRow ).toEqual( expected );
		} );

		test( 'should match the visible rows', () => {
			const { getByRole, getAllByRole } = render( <DepositsList /> );
			getByRole( 'button', { name: 'Download' } ).click();

			const csvContent = downloadCSVFile.mock.calls[ 0 ][ 1 ];
			const csvRows = csvContent.split( os.EOL );
			const displayRows = getAllByRole( 'row' );

			expect( csvRows.length ).toEqual( displayRows.length );

			const csvFirstDeposit = csvRows[ 1 ].split( ',' );
			const displayFirstDeposit = Array.from(
				displayRows[ 1 ].querySelectorAll( 'td' )
			).map( ( td ) => td.textContent );

			// Note:
			//
			// 1. CSV and display indexes are off by 1 because the first field in CSV is deposit id,
			//    which is missing in display.
			//
			// 2. The indexOf check in amount's expect is because the amount in CSV may not contain
			//    trailing zeros as in the display amount.
			//
			expect( formatDate( csvFirstDeposit[ 1 ] ) ).toBe(
				displayFirstDeposit[ 0 ]
			); // date
			expect( csvFirstDeposit[ 2 ] ).toBe( displayFirstDeposit[ 1 ] ); // type
			expect(
				getUnformattedAmount( displayFirstDeposit[ 2 ] ).indexOf(
					csvFirstDeposit[ 3 ]
				)
			).not.toBe( -1 ); // amount
			expect( csvFirstDeposit[ 4 ] ).toBe( displayFirstDeposit[ 3 ] ); // status
			expect( csvFirstDeposit[ 5 ] ).toBe(
				`"${ displayFirstDeposit[ 4 ] }"`
			); // bank account
		} );
	} );
} );

function getUnformattedAmount( formattedAmount ) {
	let amount = formattedAmount.replace( /[^0-9,.' ]/g, '' ).trim();
	amount = amount.replace( ',', '.' ); // Euro fix
	return amount;
}

function formatDate( date ) {
	return dateI18n( 'M j, Y', moment.utc( date ).toISOString(), true );
}
