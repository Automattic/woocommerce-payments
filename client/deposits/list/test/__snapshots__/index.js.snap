// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deposits list renders correctly 1`] = `
<div>
  <div
    class="woocommerce-card woocommerce-table has-menu"
  >
    <div
      class="woocommerce-card__header"
    >
      <div
        class="woocommerce-card__title-wrapper"
      >
        <h2
          class="woocommerce-card__title woocommerce-card__header-item"
        >
          Deposit history
        </h2>
      </div>
      <div
        class="woocommerce-card__menu woocommerce-card__header-item"
      >
        <div
          class="woocommerce-ellipsis-menu"
        >
          <div>
            <button
              aria-expanded="false"
              class="components-button components-icon-button woocommerce-ellipsis-menu__toggle"
              title="Choose which values to display"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="dashicon dashicons-ellipsis"
                focusable="false"
                height="20"
                role="img"
                viewBox="0 0 20 20"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 10c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm12-2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-7 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="woocommerce-card__body"
    >
      <div
        aria-hidden="false"
        aria-labelledby="caption-0"
        class="woocommerce-table__table"
        role="group"
      >
        <table>
          <caption
            class="woocommerce-table__caption screen-reader-text"
            id="caption-0"
          >
            Deposit history
          </caption>
          <tbody>
            <tr>
              <th
                class="woocommerce-table__header info-button"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                />
              </th>
              <th
                class="woocommerce-table__header date-time is-left-aligned"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Date
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Type
                </span>
              </th>
              <th
                class="woocommerce-table__header is-numeric"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Amount
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Status
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Bank account
                </span>
              </th>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item info-button"
                scope="row"
              >
                <a
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                >
                  <svg
                    class="gridicon gridicons-info-outline needs-offset"
                    height="18"
                    viewBox="0 0 24 24"
                    width="18"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M13 9h-2V7h2v2zm0 2h-2v6h2v-6zm-1-7c-4.41 0-8 3.59-8 8s3.59 8 8 8 8-3.59 8-8-3.59-8-8-8m0-2c5.523 0 10 4.477 10 10s-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2z"
                      />
                    </g>
                  </svg>
                </a>
              </th>
              <td
                class="woocommerce-table__item date-time is-left-aligned"
              >
                <a
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                >
                  Jan 2, 2020
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                  tabindex="-1"
                >
                  Deposit
                </a>
              </td>
              <td
                class="woocommerce-table__item is-numeric"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                  tabindex="-1"
                >
                  $20.00
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                  tabindex="-1"
                >
                  Paid
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock1"
                  tabindex="-1"
                >
                  MOCK BANK •••• 1234 (USD)
                </a>
              </td>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item info-button"
                scope="row"
              >
                <a
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                >
                  <svg
                    class="gridicon gridicons-info-outline needs-offset"
                    height="18"
                    viewBox="0 0 24 24"
                    width="18"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M13 9h-2V7h2v2zm0 2h-2v6h2v-6zm-1-7c-4.41 0-8 3.59-8 8s3.59 8 8 8 8-3.59 8-8-3.59-8-8-8m0-2c5.523 0 10 4.477 10 10s-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2z"
                      />
                    </g>
                  </svg>
                </a>
              </th>
              <td
                class="woocommerce-table__item date-time is-left-aligned"
              >
                <a
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                >
                  Jan 3, 2020
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                  tabindex="-1"
                >
                  Withdrawal
                </a>
              </td>
              <td
                class="woocommerce-table__item is-numeric"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                  tabindex="-1"
                >
                  $30.00
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                  tabindex="-1"
                >
                  Pending
                </a>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <a
                  class="woocommerce-table__clickable-cell"
                  data-link-type="wc-admin"
                  href="admin.php?page=wc-admin&path=%2Fpayments%2Fdeposits%2Fdetails&id=po_mock2"
                  tabindex="-1"
                >
                  MOCK BANK •••• 1234 (USD)
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
